orchestra/todos/filters.py

```diff
@@ -0,0 +1,58 @@
+import json
+
+from rest_framework import filters
+
+
+class QueryParamsFilterBackend(filters.BaseFilterBackend):
+    """
+    Takes a json-encoded data as a queryparam from a URL,
+    serializes into a python data structure
+    and passes it as a queryset arguments.
+    Format: ?q='{"key": ["val1", "val2"], "key2": null}', where `q` is
+    filter_prefix which can be set in the view:
+    filter_prefix = 'custom_key'
+    Note: this doesn't support nested__fields. To support them,
+    add `filterset_fields` iterable to the view.
+    Note: nested__field lookup in JSONField is not supported even if added
+    to the `filterset_fields`.
+    This issue can be fixed when we migrate to Django 3.1
+    and convert additional_data from django-jsonfields to the native one.
+    """
+    def _get_query_params_prefix(self, view):
+        """
+        By default, filter params are assigned to `q` parameter.
+        """
+        if hasattr(view, 'filter_prefix'):
+            return view.filter_prefix
+        return 'q'
+
+    def _get_filter_kwargs(self, view, params):
+        serializer = view.get_serializer(data=params)
+        serializer.is_valid()
+        return serializer.data
+
+    def _get_filterset_fields_kwargs(self, view, params, qs_kwargs):
+        filterset_kwargs = {}
+        if hasattr(view, 'filterset_fields'):
+            for field in view.filterset_fields:
+                if field in params and field not in qs_kwargs:
+                    filterset_kwargs[field] = params[field]
+        return filterset_kwargs
+
+    def _get_params(self, request, view):
+        """
+        Here we get JSON encoded params and encode them into Python objects.
+        """
+        params = request.query_params.dict()
+        json_params = params.pop(self._get_query_params_prefix(view), '{}')
```

Comment:
is it considered good practice to modify the query params? as in, should this `pop` be a `get`?

Reply:
Addressed :) Now we mutate a copy of the object.

Reply:
Curiously, why do we want to incur a copy when we can use `get` and leave the original unpurturbed?

Reply:
I just realized, that the `pop` version was better :)
Suppose, we passed `?project__id=8485&q={"title": "some title"}`
So in params we will have `{'project__id': '8485', 'q': '{"title": "some title"}'}`
When we **get** `q`, we leave the original 'q' key with its unconverted value and we end up with `{'project__id': '8485', 'q': '{"title": "some title"}', 'title': 'some title'} after converting things into dict and merging back.

Reply:
So, my suggestion is we
a) clone `request.query_params.copy().dict()` pop `q`, convert and merge the converted dict into params
b) clone `request.query_params.dict().copy()` pop `q`, convert and merge the converted dict into params
c) we don't touch `request.query_params.dict()`, and create a third dictionary where we merge `params` and `converted`. But we'll end up with `q` there:
        params = request.query_params.dict()
        json_params = params.get(self._get_query_params_prefix(view), '{}')
        converted = json.loads(json_params)
        result = params.copy()
        result.update(converted)
        return result